src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Hover/DefaultRazorHoverInfoService.cs

```diff
@@ -0,0 +1,167 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Composition;
+using System.Diagnostics;
+using System.Linq;
+using Microsoft.AspNetCore.Razor.Language;
+using Microsoft.AspNetCore.Razor.Language.Legacy;
+using Microsoft.AspNetCore.Razor.LanguageServer.Completion;
+using Microsoft.VisualStudio.Editor.Razor;
+using OmniSharp.Extensions.LanguageServer.Protocol.Models;
+using HoverModel = OmniSharp.Extensions.LanguageServer.Protocol.Models.Hover;
+using RangeModel = OmniSharp.Extensions.LanguageServer.Protocol.Models.Range;
+using Microsoft.AspNetCore.Razor.Language.Syntax;
+
+namespace Microsoft.AspNetCore.Razor.LanguageServer.Hover
+{
+    internal class DefaultRazorHoverInfoService : RazorHoverInfoService
+    {
+        private readonly TagHelperFactsService _tagHelperFactsService;
+        private readonly TagHelperDescriptionFactory _tagHelperDescriptionFactory;
+        private readonly HtmlFactsService _htmlFactsService;
+
+        [ImportingConstructor]
+        public DefaultRazorHoverInfoService(
+            TagHelperFactsService tagHelperFactsService,
+            TagHelperDescriptionFactory tagHelperDescriptionFactory,
+            HtmlFactsService htmlFactsService)
+        {
+            if (tagHelperFactsService is null)
+            {
+                throw new ArgumentNullException(nameof(tagHelperFactsService));
+            }
+
+            if (tagHelperDescriptionFactory is null)
+            {
+                throw new ArgumentNullException(nameof(tagHelperDescriptionFactory));
+            }
+
+            if (htmlFactsService is null)
+            {
+                throw new ArgumentNullException(nameof(htmlFactsService));
+            }
+
+            _tagHelperFactsService = tagHelperFactsService;
+            _tagHelperDescriptionFactory = tagHelperDescriptionFactory;
+            _htmlFactsService = htmlFactsService;
+        }
+
+        public override HoverModel GetHoverInfo(RazorCodeDocument codeDocument, SourceSpan location)
```

Comment:
You should add C# level tests to verify expected output on `GetHoverInfo` on its own. We do that already for all of the various completion tests. I.e. https://github.com/aspnet/AspNetCore-Tooling/blob/master/src/Razor/test/Microsoft.CodeAnalysis.Razor.Workspaces.Test/DefaultTagHelperCompletionServiceTest.cs

By adding the unit tests for the C# side of the world it means we can isolate functionality breaks without having to run functional tests and can ensure that when we're making changes to just the language server that we don't break this piece indirectly.

Reply:
I wonder how hard it would be to set up CodeCoverage in this repo...

Reply:
Oo interesting idea. Do we have that in AspNetCore?